File: frontend_flutter/test/data/settings_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:score/data/local_store.dart';
import 'package:score/data/settings.dart';
import 'package:score/notation/score_sheet.dart';

/// What the device remembers being told.
///
/// A player sets this once, in a room, and expects to find it that way at the
/// next rehearsal. That is the whole of what it has to do — but it is read
/// before the first frame, so what happens when the stored value is nonsense
/// matters more than it looks: the answer has to be a running app.

void main() {
  late LocalStore store;

  setUp(() async => store = await LocalStore.inMemory());

  test('a device that has never been asked follows the system', () async {
    final settings = await Settings.load(store);

    expect(settings.themeMode, ThemeMode.system);
  });

  test('what was chosen is what is found at the next start', () async {
    await (await Settings.load(store)).setThemeMode(ThemeMode.dark);

    // A different Settings on the same store: the app, started again.
    expect((await Settings.load(store)).themeMode, ThemeMode.dark);
  });

  test('going back to the system stores nothing rather than the word for it',
      () async {
    final settings = await Settings.load(store);
    await settings.setThemeMode(ThemeMode.light);
    await settings.setThemeMode(ThemeMode.system);

    expect(await store.readSetting('theme_mode'), isNull);
    expect((await Settings.load(store)).themeMode, ThemeMode.system);
  });

  test('the app is told, so the theme can change under it', () async {
    final settings = await Settings.load(store);
    var told = 0;
    settings.addListener(() => told++);

    await settings.setThemeMode(ThemeMode.dark);
    expect(told, 1);

    // Choosing what is already chosen is not a change to redraw for.
    await settings.setThemeMode(ThemeMode.dark);
    expect(told, 1);
  });

  test('a stored value nobody can read is the system, not a broken app',
      () async {
    await store.writeSetting('theme_mode', 'midnight');

    expect((await Settings.load(store)).themeMode, ThemeMode.system);
  });

  group('the lamp on the page', () {
    test('starts full up in the light and turned down in the dark', () async {
      final settings = await Settings.load(store);

      expect(settings.pageLook(Brightness.light).brightness, SheetPalette.full);
      expect(settings.pageLook(Brightness.dark).brightness, SheetPalette.night);
      expect(settings.pageLook(Brightness.light).warmth, 0);
    });

    test('is set for one page without touching the other', () async {
      final settings = await Settings.load(store);

      await settings.setPageLook(
          Brightness.dark, (brightness: 0.4, warmth: 0.5));

      expect(settings.pageLook(Brightness.dark).brightness, 0.4);
      expect(settings.pageLook(Brightness.light).brightness, SheetPalette.full,
          reason: 'a lit desk and a dark stage are two different rooms');
    });

    test('is where it was left at the next start', () async {
      await (await Settings.load(store))
          .setPageLook(Brightness.dark, (brightness: 0.22, warmth: 0.8));

      final reopened = await Settings.load(store);
      expect(reopened.pageLook(Brightness.dark).brightness, closeTo(0.22, 1e-9));
      expect(reopened.pageLook(Brightness.dark).warmth, closeTo(0.8, 1e-9));
    });

    test('cannot be turned below what can be read off', () async {
      final settings = await Settings.load(store);

      await settings.setPageLook(
          Brightness.dark, (brightness: 0.0, warmth: 3));

      expect(settings.pageLook(Brightness.dark).brightness,
          SheetPalette.dimmest);
      expect(settings.pageLook(Brightness.dark).warmth, 1);
    });

    test('put back is stored as nothing, the way it started', () async {
      final settings = await Settings.load(store);
      await settings.setPageLook(
          Brightness.dark, (brightness: 0.5, warmth: 0.5));
      expect(settings.isPageLookDefault(Brightness.dark), isFalse);

      await settings.resetPageLook(Brightness.dark);

      expect(settings.isPageLookDefault(Brightness.dark), isTrue);
      expect(await store.readSetting('page_look_dark'), isNull);
    });

    test('a stored page nobody can read is the page nobody has touched',
        () async {
      await store.writeSetting('page_look_dark', 'dim-ish');

      final settings = await Settings.load(store);
      expect(settings.pageLook(Brightness.dark).brightness, SheetPalette.night);
    });

    test('the page is told, so the sheet can keep up with the thumb', () async {
      final settings = await Settings.load(store);
      var told = 0;
      settings.addListener(() => told++);

      await settings.setPageLook(
          Brightness.dark, (brightness: 0.4, warmth: 0.3));
      expect(told, 1);

      // Landing on the value it already had is not a change to redraw for.
      await settings.setPageLook(
          Brightness.dark, (brightness: 0.4, warmth: 0.3));
      expect(told, 1);
    });
  });
}

File: frontend/src/domains/settings/settings.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import {defaultPageLook, DIMMEST, FULL, NIGHT, NIGHT_WARMTH} from './sheet-palette.js';

// The settings read the localStorage global at call time, so a fake put here
// before importing is enough to drive them. Which way round the machine says it
// is, is a media query, and is faked the same way.
const store = new Map();
globalThis.localStorage = {
  setItem: (key, value) => store.set(key, String(value)),
  getItem: (key) => (store.has(key) ? store.get(key) : null),
  removeItem: (key) => store.delete(key),
};

let machineIsDark = false;
globalThis.matchMedia = (query) => ({
  matches: query.includes('dark') && machineIsDark,
  addEventListener: () => undefined,
});

const {Settings, themeModeStorageKey} = await import('./settings.js');

test.beforeEach(() => {
  store.clear();
  machineIsDark = false;
});

test('a device that has never been asked follows the machine', () => {
  assert.equal(Settings.themeMode, 'system');
});

test('a way round survives a round trip', () => {
  Settings.themeMode = 'dark';

  assert.equal(Settings.themeMode, 'dark');
});

// Following the system is what an app does when it has been told nothing, so it
// is stored as nothing: a device that has never been asked and one that has
// been put back to following the machine are the same device.
test('going back to following the machine leaves nothing behind', () => {
  Settings.themeMode = 'dark';

  Settings.themeMode = 'system';

  assert.equal(store.get(themeModeStorageKey), undefined);
  assert.equal(Settings.themeMode, 'system');
});

// A key from a version that wrote them differently is not worth a page that
// will not start.
test('a way round nobody can read follows the machine', () => {
  store.set(themeModeStorageKey, 'sepia');

  assert.equal(Settings.themeMode, 'system');
});

test('what is on the screen is what was chosen, whatever the machine says', () => {
  machineIsDark = true;
  Settings.themeMode = 'light';

  assert.equal(Settings.brightness, 'light');
});

test('what is on the screen is what the machine says when nothing was chosen', () => {
  machineIsDark = true;

  assert.equal(Settings.brightness, 'dark');

  machineIsDark = false;

  assert.equal(Settings.brightness, 'light');
});

test('an untouched page is the one nobody has set', () => {
  assert.deepEqual(Settings.pageLook('light'), defaultPageLook('light'));
  assert.deepEqual(Settings.pageLook('dark'), defaultPageLook('dark'));
  assert.ok(Settings.isPageLookDefault('dark'));
});

test('a page survives a round trip', () => {
  Settings.setPageLook('dark', {brightness: 0.4, warmth: 0.5});

  assert.deepEqual(Settings.pageLook('dark'), {brightness: 0.4, warmth: 0.5});
  assert.ok(!Settings.isPageLookDefault('dark'));
});

// Two different rooms: the lamp a reader wants at a lit desk is not the one
// they want at a gig, and setting one must not set the other.
test('the light page and the dark page are set apart from one another', () => {
  Settings.setPageLook('dark', {brightness: 0.4, warmth: 0.5});

  assert.deepEqual(Settings.pageLook('light'), defaultPageLook('light'));
});

test('a page set past either end is held at the end', () => {
  Settings.setPageLook('dark', {brightness: 0, warmth: 3});

  assert.deepEqual(Settings.pageLook('dark'), {brightness: DIMMEST, warmth: 1});
});

test('putting a page back leaves nothing behind', () => {
  Settings.setPageLook('dark', {brightness: 0.4, warmth: 0.5});

  Settings.resetPageLook('dark');

  assert.equal(store.get('score-page-look-dark'), undefined);
  assert.deepEqual(Settings.pageLook('dark'), {brightness: NIGHT, warmth: NIGHT_WARMTH});
  assert.ok(Settings.isPageLookDefault('dark'));
});

// Half of a pair of numbers, or none, is the page nobody has touched rather
// than a page that cannot be drawn.
test('a page nobody can read is the page nobody has touched', () => {
  for (const stored of ['', '0.4', 'a,b', '0.4,0.5,0.6', 'null']) {
    store.set('score-page-look-light', stored);

    assert.deepEqual(Settings.pageLook('light'), {brightness: FULL, warmth: 0},
      `"${stored}" should have read as an untouched page`);
  }
});

test('the palette that is applied is the one for the room the app is in', () => {
  machineIsDark = true;
  Settings.setPageLook('dark', {brightness: FULL, warmth: 0});

  assert.equal(Settings.sheetPalette.paper, '#ffffff');
});

test('applying puts the way round and the page on the document', () => {
  Settings.themeMode = 'dark';
  Settings.setPageLook('dark', {brightness: FULL, warmth: 0});
  const root = _fakeRoot();

  Settings.apply(root);

  assert.equal(root.attributes.get('data-theme'), 'dark');
  assert.equal(root.style.properties.get('--paper'), '#ffffff');
  assert.equal(root.style.properties.get('--ink'), '#000000');
});

test('following the machine is said by saying nothing', () => {
  const root = _fakeRoot();
  root.attributes.set('data-theme', 'dark');
  Settings.themeMode = 'system';

  Settings.apply(root);

  assert.equal(root.attributes.has('data-theme'), false);
});

/**
 * As much of an element as applying the settings touches.
 */
function _fakeRoot() {
  const attributes = new Map();
  const properties = new Map();
  return {
    attributes,
    style: {properties, setProperty: (name, value) => properties.set(name, value)},
    setAttribute: (name, value) => attributes.set(name, value),
    removeAttribute: (name) => attributes.delete(name),
  };
}
